src/common/locallaplaciancl.c

```diff
@@ -130,58 +130,40 @@ cl_int dt_local_laplacian_cl(
 
   if(b->bwidth <= 1 || b->bheight <= 1) return err;
 
-  size_t sizes_pad[] = { ROUNDUPDWD(b->bwidth, b->devid), ROUNDUPDHT(b->bheight, b->devid), 1 };
-  dt_opencl_set_kernel_arg(b->devid, b->global->kernel_pad_input, 0, sizeof(cl_mem), &input);
-  dt_opencl_set_kernel_arg(b->devid, b->global->kernel_pad_input, 1, sizeof(cl_mem), &b->dev_padded[0]);
-  dt_opencl_set_kernel_arg(b->devid, b->global->kernel_pad_input, 2, sizeof(int), &b->width);
-  dt_opencl_set_kernel_arg(b->devid, b->global->kernel_pad_input, 3, sizeof(int), &b->height);
-  dt_opencl_set_kernel_arg(b->devid, b->global->kernel_pad_input, 4, sizeof(int), &b->max_supp);
-  dt_opencl_set_kernel_arg(b->devid, b->global->kernel_pad_input, 5, sizeof(int), &b->bwidth);
-  dt_opencl_set_kernel_arg(b->devid, b->global->kernel_pad_input, 6, sizeof(int), &b->bheight);
-  err = dt_opencl_enqueue_kernel_2d(b->devid, b->global->kernel_pad_input, sizes_pad);
+  err = dt_opencl_enqueue_kernel_2d_args(b->devid, b->global->kernel_pad_input, b->bwidth, b->bheight,
+    CLARG(input), CLARG(b->dev_padded[0]), CLARG(b->width), CLARG(b->height), CLARG(b->max_supp), CLARG(b->bwidth),
+    CLARG(b->bheight));
   if(err != CL_SUCCESS) goto error;
 
   // create gauss pyramid of padded input, write coarse directly to output
   for(int l=1;l<b->num_levels;l++)
   {
     const int wd = dl(b->bwidth, l), ht = dl(b->bheight, l);
     size_t sizes[] = { ROUNDUPDWD(wd, b->devid), ROUNDUPDHT(ht, b->devid), 1 };
-    dt_opencl_set_kernel_arg(b->devid, b->global->kernel_gauss_reduce, 0, sizeof(cl_mem), &b->dev_padded[l-1]);
+    dt_opencl_set_kernel_args(b->devid, b->global->kernel_gauss_reduce, 0, CLARG(b->dev_padded[l-1]));
     if(l == b->num_levels-1)
-      dt_opencl_set_kernel_arg(b->devid, b->global->kernel_gauss_reduce, 1, sizeof(cl_mem), &b->dev_output[l]);
+      dt_opencl_set_kernel_args(b->devid, b->global->kernel_gauss_reduce, 1, CLARG(b->dev_output[l]));
     else
-      dt_opencl_set_kernel_arg(b->devid, b->global->kernel_gauss_reduce, 1, sizeof(cl_mem), &b->dev_padded[l]);
-    dt_opencl_set_kernel_arg(b->devid, b->global->kernel_gauss_reduce, 2, sizeof(int), &wd);
-    dt_opencl_set_kernel_arg(b->devid, b->global->kernel_gauss_reduce, 3, sizeof(int), &ht);
+      dt_opencl_set_kernel_args(b->devid, b->global->kernel_gauss_reduce, 1, CLARG(b->dev_padded[l]));
+    dt_opencl_set_kernel_args(b->devid, b->global->kernel_gauss_reduce, 2, CLARG(wd), CLARG(ht));
     err = dt_opencl_enqueue_kernel_2d(b->devid, b->global->kernel_gauss_reduce, sizes);
     if(err != CL_SUCCESS) goto error;
   }
 
   for(int k=0;k<num_gamma;k++)
   { // process images
     const float g = (k+.5f)/(float)num_gamma;
-    dt_opencl_set_kernel_arg(b->devid, b->global->kernel_process_curve, 0, sizeof(cl_mem), &b->dev_padded[0]);
```

Comment:
This one should also be revert to avoid the issue.

That being said, I have no idea why at the moment.

Reply:
I have an untested theory that the problem is the last two arguments, b->bwidth and b->bheight. They are declared as `size_t`, but the `dt_opencl_set_kernel_arg` calls used to think they were `sizeof(int)` (=4) long. The new code ("correctly") sends their size as `sizeof(size_t)` =8, so maybe its not surprising that things fall over.

@jenshannoschwalm would there have been a good reason for the mismatch between variable sent and its claimed size? I suppose this could be fixed on the kernel side by correctly declaring the parameter as size_t (if that type is supported in cl) or in the definition of `dt_local_laplacian_cl_t`?

Maybe I'll be able to produce a dump of all variable sizes sent for all kernels in both before and after versions and if they line up that might be an easy way to check all in one go. Currently I'm trying to get pocl or oclgrind to work on my development machine (that doesn't have a GPU) so it is easier for me to test these things.

Reply:
I will have to look that up but from memory opencl size_t might be  not safe.

Reply:
Would be cl_long or cl_ulong

Reply:
I am just on mobile but indeed the size_t for pad looks wrong. The kernel seems to take int.

Reply:
size_t seems awfully big for one dimension of an image, but I guess the advantage is you can safely multiply without having to cast first. If I remember correctly @johnny-bit did a lot of work on that at some point, but enforcing consistency, where _all_ `width`s and `height`s would be size_t is a long way off. So then maybe having them all be ints and enforcing casts everywhere is the more consistent way? Or just deal with them case by case. Copy/pastes from a situation using size_t to one with ints will continue to introduce problems though.

At least I feel that this PR making this explicit, rather than just accepting mismatched `sizeof(type)`s and actual `type`s is "better"?

Reply:
Size_t is simply not required for image operations in CL code. Int as coordinates are fine.